New package: 3D Alpha Wrapping (#6432)
CGAL/cgal: Alpha_wrap_3/doc/Alpha_wrap_3/alpha_wrap_3.txt

namespace CGAL {

/*!

\mainpage User Manual
\anchor Chapter_3D_Alpha_wrapping
\cgalAutoToc

\authors TBA

<center>
<img src="aw3_bike_lod.png" style="max-width:70%;"/>
</center>

\section aw3_introduction Introduction

Various tasks in geometric modeling and processing require 3D objects represented as valid surface meshes,
where "valid" refers to meshes that are watertight, intersection-free, orientable, and 2-manifold.
Such representations offer well-defined notions of interior/exterior and geodesic neighborhoods.

3D data are usually acquired through measurements followed by reconstruction, designed by humans,
or generated through imperfect automated processes.
As a result, they can exhibit a wide variety of defects including gaps, missing data,

> self-intersections, degeneracies such as zero-volume structures, and non-manifold features.

Reviewer: ```suggestion
self-intersections, degeneracies, and non-manifold features.
```

to repair specific defects, usually with the goal of guaranteeing specific properties in the repaired 3D model.
Reliably repairing all types of defects is notoriously difficult and is often an ill-posed problem
as many valid solutions exist for a given 3D model with defects.
In addition, the input model can be overly complex with unnecessary geometric details,
spurious topological structures, nonessential inner components, or excessively fine discretizations.
For applications such as collision avoidance, path planning, or simulation,
getting an approximation of the input can be more relevant than repairing it.
Approximation herein refers to an approach capable of filtering out inner structures,
fine details and cavities, as well as wrapping the input within a user-defined offset margin.

Given an input 3D geometry, we address the problem of computing a conservative approximation,
where conservative means that the output is guaranteed to strictly enclose the input.
We seek unconditional robustness in the sense that the output mesh should be valid (oriented,
2-manifold, and without self-intersections), even for raw input with many defects
and degeneracies.

> The default input is a soup of 3D triangles, but the generic interface leaves the door open

Reviewer: should be updated as points and segments are handled, right?

Author: Segments API is fully undocumented and I don't want to put emphasis on the Points interface.

to other types of finite 3D primitives.

\cgalFigureAnchor{1}
<center>
<img src="aw3_triangle_soup.png" style="max-width:70%;"/>
</center>
\cgalFigureCaptionBegin{1}
Shrink-wrapping output from a triangle soup, with many intersections and gaps.
From left to right, input model, output wrap, and superposition.
\cgalFigureCaptionEnd

\cgalFigureAnchor{2}
<center>
<img src="aw3_non_manifold_cases.png" style="max-width:75%;"/>
</center>
\cgalFigureCaptionBegin{2}
Dealing with non-manifold features and degeneracies.
From left to right, a non-manifold vertex, self-intersecting faces and two adjacent triangles
representing a zero-volume structure.
The algorithm handles these cases by wrapping an offset of the input.
\cgalFigureCaptionEnd

> \section aw3_definition Approach

Reviewer: In general, we place the theory section in the last section of the doc to start with how to use the code.

> 

Reviewer: Either move that section to the end or add a sentence linking to the C++ user API section.

Reviewer: OK it's not a big deal here as the API section is not essential anyway + the math helps with alpha/offset. Marking it as resolved.

new alpha-traversable gates are pushed to the queue, and traversal is resumed.
Grey edges depict the Delaunay triangulation. Blue edges depict the Voronoi diagram.
Pink circles depict the empty circle of radius alpha. The output edges (dark blue) separate inside from outside triangles.
\cgalFigureCaptionEnd

\subsection aw3_guarantees Guarantees

The algorithm is proven to terminate and to produce a 2-manifold triangulated surface mesh
that strictly encloses the input data.
The key element to the proof is that we wrap from outside to inside and never allow a cell
that intersects the input to be flagged inside.
Furthermore, both criteria that lead to refinement of the triangulation insert Steiner points
that are guaranteed to break the cells in need of refinement and reduce the neighbor facets circumradii.

Because the main refinement criterion is the insertion of an intersection between a dual Voronoi edge
and an offset of the input, or the projection of a Voronoi vertex onto the offset of the input,
the algorithm has similarities to popular meshing algorithms based on Delaunay filtering
and refinement (see \ref Chapter_3D_Mesh_Generation).

\section aw3_interface Interface

> Our algorithm takes as input a set of triangles in 3D, provided either as a triangle soup or

Reviewer: only triangles?

Author: See above.

as a triangle surface mesh, and two user-defined scalar parameters: the <em>alpha</em> and the <em>offset</em> values.
It proceeds by shrink-wrapping and refining a 3D Delaunay triangulation starting from a loose bounding box of the input.
The parameter <em>alpha</em> refers to the size of cavities or holes that cannot be traversed during wrapping,
and hence to the final level of detail, as alpha acts like a sizing field in a common Delaunay
refinement algorithm (\ref Chapter_3D_Mesh_Generation).
The parameter <em>offset</em> refers to the distance between the vertices of the refined triangulation
and the input, so that a large offset translates into a loose enclosing of the input.
This second parameter offers a means to control the trade-off between tightness and complexity.

> The main entry point of the component is a global function that generates the alpha wrap;

Reviewer: ```suggestion
The main entry point of the component is the global function `CGAL::alpha_wrap_3()` that generates the alpha wrap;
```

> this function takes as input a polygon soup or a polygon mesh: `CGAL::alpha_wrap_3()`.

Reviewer: ```suggestion
this function takes as input a polygon soup or a polygon mesh.
```

There is no prerequisite on the input connectivity so that it can take an arbitrary triangle soup,

> with islands, self-intersections, or overlaps, as well as combinatorial or geometrical degeneracies.

Reviewer: is islands clearly defined here? Mentioning closeness and connected components might not be either as you mention a soup.

Author: I think hand waving like this is clear enough given that we are not actually going to give an algorithm that aims to perform on specific configurations that would require more precise definitions.

> The underlying traits class must be a model of the `AlphaWrapTraits_3` concept. It should use

Reviewer: No longer exists

\cgalFigureCaptionEnd

\subsection aw3_two_side A Note on "Two-Sided" Wraps

The offset parameter is crucial to our approach because it guarantees that the output is a closed,
2-manifold surface mesh.
Indeed, and even when the input is a zero-volume structure such as a single 3D triangle,
the output wrap is a thin volume enclosing the said triangle \cgalFigureRef{2}.

Users should keep in mind that the wrapping algorithm has no means of determining whether it is acting on the inside or the outside
of the unsigned distance field, and will thus produce two-sided wraps in the case of holes in the input
and values of alpha smaller than the size of the holes.

\cgalFigureAnchor{9}
<center>
<img src="aw3_double_sided.png" style="max-width:80%;"/>
</center>
\cgalFigureCaptionBegin{9}
Two-sided wrap.
(Left) Wrapping a Bunny in 2D, with decreasing values for alpha.
(Right) Wrapping a defect-laden Bunny in 3D. The rightmost column depicts a clipped visualization
of the inside. When alpha is small enough with respect the diamater of the holes, the algorithm generates a two-sided wrap.
\cgalFigureCaptionEnd

\section aw3_performance Performance

> The charts below plots the computation times of the wrapping algorithm on the Thingi10k dataset, as well as the complexity of the output triangle mesh.

Reviewer: ref?
